JumpscaleLibs/sal/zosv2/Zosv2.py

```diff
@@ -0,0 +1,197 @@
+        self.config = NetworkConfig()
+
+    # consider a node up if it has received update during the last 10 minute
+    def is_up(self, node):
+        ago = j.data.time.epoch - (60 * 10)
+        return node.updated > ago
+
+    def get_node(self,ip_version):
+        #return node and public ip
+        nodes = self.explorer.actors_all.nodes.list().nodes
+        return self.find_node_public(nodes,ip_version)
+
+    def find_node_public(self, nodes, version=6):
+        # search a node that has a public ipv4 address
+        for node in filter(self.is_up, nodes):
+            for iface in node.ifaces:
+                for addr in iface.addrs:
+                    ip = ipaddress.ip_interface(addr).ip
+                    if ip.version != version:
+                        continue
+                    if ip.is_global:
+                        self.config.public_ip = str(ip)
+                        return (node, str(ip))
+        if version==4:
+            for node in filter(self.is_up, nodes):
+                ip = ipaddress.ip_interface(node.public_config.ipv4).ip
+                if ip.is_global:
+                    self.config.public_ip = str(ip)
+                    return (node, str(ip))
+            # need to check in ifaces
+    def find_free_wg_port(self, node):
+        ports = set(list(range(6000, 9000)))
+        used = set(node.wg_ports)
+        free = ports - used
+        return free.pop()
+
+    def new_reservation(self, user):
+        reservation_model = j.data.schema.get_from_url("tfgrid.workloads.reservation.1")
+        reservation = reservation_model.new()
+        reservation.customer_tid = user.id
+        reservation.data_reservation.expiration_provisioning = j.data.time.epoch + (60 * 10)  # 10 minutes
+        reservation.data_reservation.expiration_reservation = j.data.time.epoch + (60 * 10)  # 10 minutes
+        return  reservation
+
+    def get_user(self,name,email):
+        # should return error if doesn't exist
+        return self.explorer.actors_all.phonebook.get(name=name, email=email)
+
+    def check_ip_in_network(self, ip , network):
+        return netaddr.IPAddress(ip) in netaddr.IPNetwork(network)
+
+    def key_pair_get(self):
+        wg_private = public.PrivateKey.generate()
+        wg_public = wg_private.public_key
+
+        wg_private_base64 = wg_private.encode(Base64Encoder)
+        wg_public_base64 = wg_public.encode(Base64Encoder)
+
+        return wg_private_base64, wg_public_base64
+
+    def create_network(self,reservation,ip_range,node_id):
+        network = reservation.data_reservation.networks.new()
+        network.node_id = node_id
+        network.workload_id = 1
+        network.name = j.data.idgenerator.generateXCharID(16)
+        network.iprange = ip_range
+        return reservation , network
+
+    def access_node(self, network, ip_range, node):
+        network_range = netaddr.IPNetwork(ip_range).ip
+        # add 256 to plus 1 in network from 172.20.0.0 -> 172.20.1.0
```

Comment:
Why do you need to modify the ip_range specify be the user ?

Reply:
i'm not sure , but i did it to get configs of Wire-guard 
example from :
https://github.com/threefoldtech/jumpscaleX_threebot/blob/development/ThreeBotPackages/tfgrid/provision_examples/ubuntu_ssh.py#L56

Reply:
After check this md file https://github.com/threefoldtech/zos/blob/master/docs/network/mesh.md#interworkings
that mean parent network is provided by the user and for each node added to this network it has to be assigned to a subnet of the parent network
parent network -> 172.20.0.0/16 ( chosen by user)
Node -> 172.20.1.0/24
host -> 172.20.2.0/24